README.md
# Genesis 404 Page

Customize the content of the 404 Page within the Genesis Framework.

## Description

Customizing the contents of the _Page Not Found_ page in any WordPress theme can be tricky, and usually involes editing the 404.php template file. The Genesis Framework already abstracts the default content of the page into a function, and this plugin can unhook that and replace it with your own custom title and content.

## Screenshots

![Screenshot showing input for custom 404 Not Found title and content](assets/screenshot-1.png)  
_Admin screen showing the fields for custom 404 Not Found page title and content._

## Requirements
 * WordPress 3.0+
 * Genesis Framework 1.8+ (supports HTML5 for Genesis Framework 2.0+)

## Installation

### WP .org repo

1. On the Add Plugin screen, search for _Genesis 404 Page_ and follow the usual steps.

### Upload

1. Download the latest tagged archive (choose the "zip" option).
2. Go to the __Plugins -> Add New__ screen and click the __Upload__ tab.
3. Upload the zipped archive directly.
4. Go to the Plugins screen and click __Activate__.

### Manual

1. Download the latest tagged archive (choose the "zip" option).
2. Unzip the archive.
3. Copy the folder to your `/wp-content/plugins/` directory.
4. Go to the Plugins screen and click __Activate__.

Check out the Codex for more information about [installing plugins manually](http://codex.wordpress.org/Managing_Plugins#Manual_Plugin_Installation).

### Git

Using git, browse to your `/wp-content/plugins/` directory and clone this repository:

`git clone git@github.com:billerickson/Genesis-404-Page.git`

Then go to your Plugins screen and click __Activate__.

## Updates

This plugin supports the [GitHub Updater](https://github.com/afragen/github-updater) plugin, so if you install that, this plugin becomes automatically updateable direct from GitHub.

See [CHANGES.md](changes.md) for the plugin changelog.

## Usage

Once installed, you can customize your 404 page's title and content in Genesis > 404 Page. 

Use the [genesis-404-search] shortcode to add a search form to the page.

If you'd like to dynamically list content (ex: most recent posts), I recommend you install the [Display Posts Shortcode](http://www.wordpress.org/extend/plugins/display-posts-shortcode/) and use it in the 404 page's content.

## Credits

Built by [Bill Erickson](http://twitter.com/BillErickson) and [Gary Jones](https://twitter.com/GaryJ)  
Copyright 2012 [Bill Erickson](http://www.billerickson.net)


CHANGES.md
# Changes to Genesis 404 Page

## 1.5.0 (2015-04-17)
* Add genesis layout support (Joshua David Nelson, @joshuadnelson)

## 1.4.0 (2013-11-29)
* Add setup for unit tests (Gary Jones).
* Add contextual help.
* Add support for GitHub Updater plugin (Gary Jones).
* Improved appearance of page title input field (Gary Jones).
* Improved loading of language files (Gary Jones).
* Improve support for activation on multi-sites (Gary Jones).
* Improve documentation and code standards (Gary Jones).
* Improve code organization by classes into their own files (Gary Jones).
* Fix incorrect plugin URL (Gary Jones).
* Fix incorrect textdomains (Gary Jones).

## 1.3.0 (2014-04-17)
* Updated editor for WordPress 3.9, props @joshuadavidnelson.

## 1.2.0 (2012-05-25)
* Add HTML5 Support (Bill Erickson).

## 1.1.0 (2013-01-11)
* Add Search functionality (Bill Erickson).

## 1.0.0 (2012-03-27)
* Initial release (Bill Erickson).


tests/bootstrap.php
<?php
/**
 * Bootstrap the plugin unit testing environment. Customize 'active_plugins'
 * setting below to point to your main plugin file.
 *
 * Requires WordPress Unit Tests (http://unit-test.svn.wordpress.org/trunk/).
 *
 * @package wordpress-plugin-tests
 */

// Add this plugin to WordPress for activation so it can be tested.

$GLOBALS['wp_tests_options'] = array(
	'active_plugins' => array( "genesis-404-page/plugin.php" ),
);

// If the wordpress-tests repo location has been customized (and specified
// with WP_TESTS_DIR), use that location. This will most commonly be the case
// when configured for use with Travis CI.

// Otherwise, we'll just assume that this plugin is installed in the WordPress
// SVN external checkout configured in the wordpress-tests repo.

if( false !== getenv( 'WP_TESTS_DIR' ) ) {
	require getenv( 'WP_TESTS_DIR' ) . '/includes/bootstrap.php';
} else {
	// Maybe I'm testing this plugin in the wrong place, but I need a different path to get from here to
	// the core bootstrap.php, than what Jetpack and other prominent plugins seem to use.
	// require dirname( dirname( dirname( dirname( dirname( dirname( __FILE__ ) ) ) ) ) ) . '/includes/bootstrap.php';
	require dirname( dirname( dirname( dirname( dirname( dirname( __FILE__ ) ) ) ) ) ) . '/tests/phpunit/includes/bootstrap.php';

}

class BE_Genesis_404_TestCase extends WP_UnitTestCase {

}


tests/test-test.php
<?php

/**
 * @group genesis-404-page
 */
class BE_Genesis_404_Test extends BE_Genesis_404_TestCase {
	function test_test() {
		$this->assertTrue( true );
	}
}


plugin.php
<?php
/**
 * Genesis 404
 *
 * @package           Genesis_404
 * @author            Bill Erickson
 * @license           GPL-2.0+
 * @link              https://github.com/billerickson/Genesis-404-Page
 * @copyright         2012 Bill Erickson
 *
 * @wordpress-plugin
 * Plugin Name:       Genesis 404
 * Plugin URI:        https://github.com/billerickson/Genesis-404-Page
 * Description:       Customize the content of your 404 (not found) page.
 * Version:           1.5.0
 * Author:            Bill Erickson
 * Author URI:        http://www.billerickson.net
 * Text Domain:       genesis-404-page
 * License:           GPL-2.0+
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
 * Domain Path:       /languages
 * GitHub Plugin URI: https://github.com/billerickson/Genesis-404-Page
 * GitHub Branch:     master
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the GNU
 * General Public License version 2, as published by the Free Software Foundation.  You may NOT assume
 * that you can use any other version of the GPL.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 */

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
	die;
}

require_once plugin_dir_path( __FILE__ ) . 'includes/class-be-genesis-404.php';
new BE_Genesis_404;

/**
 * Register an admin page for this plugin.
 *
 * @since 1.0.0
 * @author Bill Erickson
 */
function be_register_genesis_404_settings() {
	require_once plugin_dir_path( __FILE__ ) . 'includes/class-be-genesis-404-settings.php';

	global $_be_genesis_404_settings;
	$_be_genesis_404_settings = new BE_Genesis_404_Settings;

}
add_action( 'genesis_admin_menu', 'be_register_genesis_404_settings'  );

// Register hooks that are fired when the plugin is activated and deactivated respectively.
register_activation_hook( __FILE__, array( 'BE_Genesis_404', 'activate' ) );
register_deactivation_hook( __FILE__, array( 'BE_Genesis_404', 'deactivate' ) );


includes/class-be-genesis-404.php
<?php
/**
 * Genesis 404
 *
 * @package   Genesis_404
 * @author    Bill Erickson
 * @license   GPL-2.0+
 * @link      https://github.com/billerickson/Genesis-404-Page
 * @copyright 2015 Bill Erickson
 */

/**
 * Plugin class.
 *
 * @package Genesis_404
 * @author  Bill Erickson
 */
class BE_Genesis_404 {

	/**
	 * Unique identifier for your plugin.
	 *
	 * Use this value (not the variable name) as the text domain when internationalizing strings of text. It should
	 * match the Text Domain file header in the main plugin file.
	 *
	 * @since 1.0.0
	 *
	 * @type string
	 */
	const PLUGIN_SLUG = 'genesis-404-page';

	/**
	 * Initialize the plugin by setting localization and new site activation hooks.
	 *
	 * @since 1.0.0
	 */
	public function __construct() {
		// Load plugin text domain
		add_action( 'init', array( $this, 'load_plugin_textdomain' ) );

		// Check to see if should be used
		add_action( 'genesis_meta', array( $this, 'maybe_custom_404' ) );

		// Search Shortcode
		add_shortcode( 'genesis-404-search', array( $this, 'search_shortcode' ) );

		// Activate plugin when new blog is added
		add_action( 'wpmu_new_blog', array( $this, 'activate_new_site' ) );
	}

	/**
	 * Fired when the plugin is activated.
	 *
	 * @since 1.3.0
	 *
	 * @param boolean $network_wide True if WPMU superadmin uses "Network Activate" action, false if WPMU is
	 *                              disabled or plugin is activated on an individual blog.
	 */
	public static function activate( $network_wide ) {
		if ( function_exists( 'is_multisite' ) && is_multisite() ) {
			if ( $network_wide  ) {
				// Get all blog ids
				$blog_ids = self::get_blog_ids();

				foreach ( $blog_ids as $blog_id ) {
					switch_to_blog( $blog_id );
					self::single_activate();
				}
				restore_current_blog();
			} else {
				self::single_activate();
			}
		} else {
			self::single_activate();
		}
	}

	/**
	 * Fired when the plugin is deactivated.
	 *
	 * @since 1.3.0
	 *
	 * @param boolean $network_wide True if WPMU superadmin uses "Network Deactivate" action, false if WPMU is
	 *                              disabled or plugin is deactivated on an individual blog.
	 */
	public static function deactivate( $network_wide ) {
		if ( function_exists( 'is_multisite' ) && is_multisite() ) {
			if ( $network_wide ) {
				// Get all blog ids
				$blog_ids = self::get_blog_ids();

				foreach ( $blog_ids as $blog_id ) {
					switch_to_blog( $blog_id );
					self::single_deactivate();
				}
				restore_current_blog();
			} else {
				self::single_deactivate();
			}
		} else {
			self::single_deactivate();
		}
	}

	/**
	 * Fired when a new site is activated with a WPMU environment.
	 *
	 * @since 1.3.0
	 *
	 * @param int     $blog_id ID of the new blog.
	 */
	public function activate_new_site( $blog_id ) {
		if ( 1 !== did_action( 'wpmu_new_blog' ) )
			return;

		switch_to_blog( $blog_id );
		self::single_activate();
		restore_current_blog();
	}

	/**
	 * Get all blog ids of blogs in the current network that are:
	 * * not archived
	 * * not spam
	 * * not deleted
	 *
	 * @since 1.3.0
	 *
	 * @return array|false The blog ids, false if no matches.
	 */
	private static function get_blog_ids() {
		global $wpdb;

		// get an array of blog ids
		$sql = "SELECT blog_id FROM $wpdb->blogs
			WHERE archived = '0' AND spam = '0'
			AND deleted = '0'";
		return $wpdb->get_col( $sql );
	}

	/**
	 * Fired for each blog when the plugin is activated.
	 *
	 * @since 1.3.0
	 */
	private static function single_activate() {
		if ( 'genesis' !== basename( TEMPLATEPATH ) ) {
			deactivate_plugins( plugin_dir_path( plugin_basename( __FILE__ ) ) . 'plugin.php' );
			wp_die( sprintf( __( 'Sorry, you can&rsquo;t activate unless you have installed <a href="%s">Genesis</a>', 'genesis-404-page' ), 'http://www.billerickson.net/get-genesis' ) );
		}
	}

	/**
	 * Fired for each blog when the plugin is deactivated.
	 *
	 * @since 1.3.0
	 */
	private static function single_deactivate() {
		// TODO: Define deactivation functionality here
	}

	/**
	 * Load the plugin text domain for translation.
	 *
	 * @since 1.3.0
	 */
	public function load_plugin_textdomain() {
		$domain = self::PLUGIN_SLUG;
		$locale = apply_filters( 'plugin_locale', get_locale(), $domain );

		load_textdomain( $domain, trailingslashit( WP_LANG_DIR ) . $domain . '/' . $domain . '-' . $locale . '.mo' );
		load_plugin_textdomain( $domain, FALSE, basename( dirname( __FILE__ ) ) . '/languages' );
	}

	/**
	 * Check if custom 404 should be used.
	 *
	 * @since 1.0.0
	 * @author Bill Erickson
	 */
	public function maybe_custom_404() {
		if( !is_404() )
			return;
		
		// Replace content
		if ( genesis_get_option( 'content', 'genesis-404' ) ) {
			remove_action( 'genesis_loop', 'genesis_404' );
			add_action( 'genesis_loop', array( $this, 'loop' ) );
		}
		
		// Set layout
		if ( genesis_get_option( 'genesis_layout', 'genesis-404' ) ) {
			add_filter( 'genesis_pre_get_option_site_layout', array( $this, 'custom_layout' ) );
		}
	}

	/**
	 * Genesis 404 Loop.
	 *
	 * Defers to markup-type specific methods.
	 *
	 * @since 1.0.0
	 * @author Bill Erickson
	 */
	public function loop() {
		$title   = genesis_get_option( 'title', 'genesis-404' );
		$content = genesis_get_option( 'content', 'genesis-404' );

		if ( function_exists( 'genesis_html5' ) && genesis_html5() ) {
			$this->loop_xhtml( $title, $content );
		} else {
			$this->loop_html5( $title, $content );
		}
	}

	/**
	 * Echo 404 loop markup as XHTML.
	 *
	 * @since 1.3.0
	 *
	 * @param  string $title   Page title.
	 * @param  string $content Page content.
	 */
	protected function loop_xhtml( $title, $content ) {
		echo '<article class="page type-page status-publish entry" itemscope="" itemtype="http://schema.org/CreativeWork">';

		if ( ! empty( $title ) ) {
			echo '<header class="entry-header"><h1 class="entry-title" itemprop="headline">' . esc_attr( $title ) . '</h1></header>';
		}

		do_action( 'genesis_404_before_content' );

		if ( ! empty( $content ) ) {
			echo '<div class="entry-content" itemprop="text">' . apply_filters( 'the_content', $content ) . '</div>';
		}

		do_action( 'genesis_404_after_content' );

		echo '</article>';
	}

	/**
	 * Echo 404 loop markup as HTML5.
	 *
	 * @since 1.3.0
	 *
	 * @param  string $title   Page title.
	 * @param  string $content Page content.
	 */
	protected function loop_html5( $title, $content ) {
		echo '<div class="post hentry">';

		if ( ! empty( $title ) ) {
			echo '<h1 class="entry-title">' . esc_attr( $title ) . '</h1>';
		}

		do_action( 'genesis_404_before_content' );

		if ( ! empty( $content ) ) {
			echo '<div class="entry-content">' . apply_filters( 'the_content', $content ) . '</div>';
		}

		do_action( 'genesis_404_after_content' );

		echo '</div>';
	}
	
	/**
	 * Set the custom genesis layout.
	 *
	 * @since 1.5.0
	 * @author Joshua David Nelson, josh@joshuadnelson.com
	 *
	 * @param string $layout The current layout.
	 *
	 * @return string $layout The modified layout.
	 */
	public function custom_layout( $layout ) {
		$layout_404 = genesis_get_option( 'genesis_layout', 'genesis-404' );
		
		// Return the original layout if 'default' is selected
		if ( $layout_404 == 'default' )
			return $layout;
		
		// Return the custom layout
		return $layout_404;
		
	}

	/**
	 * Search shortcode.
	 *
	 * @since 1.1.0
	 * @author Bill Erickson
	 */
	public function search_shortcode() {
		return '<div class="genesis-404-search">' . get_search_form( false ) . '</div>';
	}

}


includes/class-be-genesis-404-settings.php
<?php
/**
 * Genesis 404
 *
 * @package   Genesis_404
 * @author    Bill Erickson
 * @license   GPL-2.0+
 * @link      https://github.com/billerickson/Genesis-404-Page
 * @copyright 2015 Bill Erickson
 */

/**
 * Admin page class.
 *
 * @package Genesis_404
 * @author  Bill Erickson
 */
class BE_Genesis_404_Settings extends Genesis_Admin_Boxes {

	/**
	 * Create an admin menu item and settings page.
	 *
	 * @since 1.0.0
	 */
	public function __construct() {
		// Specify a unique page ID.
		$page_id = 'genesis-404';

		// Set it as a child to genesis, and define the menu and page titles
		$menu_ops = array(
			'submenu' => array(
				'parent_slug' => 'genesis',
				'page_title'  => __( 'Genesis 404 Page', 'genesis-404-page' ),
				'menu_title'  => __( '404 Page', 'genesis-404-page' ),
			)
		);

		// Set up page options. These are optional, so only uncomment if you want to change the defaults
		$page_ops = array(
			// 'screen_icon'       => 'options-general',
			// 'save_button_text'  => 'Save Settings',
			// 'reset_button_text' => 'Reset Settings',
			// 'save_notice_text'  => 'Settings saved.',
			// 'reset_notice_text' => 'Settings reset.',
		);

		// Give it a unique settings field.
		// You'll access them from genesis_get_option( 'option_name', 'child-settings' );
		$settings_field = 'genesis-404';

		// Set the default values
		$default_settings = array(
			'title'   => '',
			'content' => '',
		);

		// Create the Admin Page
		$this->create( $page_id, $menu_ops, $page_ops, $settings_field, $default_settings );

		// Initialize the Sanitization Filter
		add_action( 'genesis_settings_sanitizer_init', array( $this, 'sanitization_filters' ) );
	}

	/**
	 * Set up sanitization filters.
	 *
	 * See genesis/lib/classes/sanitization.php for all available filters.
	 *
	 * @since 1.0.0
	 */
	public function sanitization_filters() {
		genesis_add_option_filter(
			'no_html',
			$this->settings_field,
			array(
				'title',
				'genesis_layout',
			)
		);

		genesis_add_option_filter(
			'safe_html',
			$this->settings_field,
			array(
				'content',
			)
		);
	}

	/**
	 * Register metaboxes on admin page.
	 *
	 * @since 1.0.0
	 */
	public function metaboxes() {

		add_meta_box( 'metabox_404', __( '404 Page', 'genesis-404-page' ), array( $this, 'metabox_404' ), $this->pagehook, 'main', 'high' );
		
		if( apply_filters( 'genesis_404_layout_box', true ) )
			add_meta_box( 'genesis_404_layout_box', __( 'Layout Settings', 'genesis' ), array( $this, 'layout_box' ), $this->pagehook, 'main', 'high' );

	}
	
	/**
	 * The layout metabox for the 404 page.
	 *
	 * @since 1.5.0
	 */
	public function layout_box() {
		$layout = esc_attr( $this->get_field_value( 'genesis_layout' ) );

		?>
		<div class="genesis-layout-selector">
			<p><input type="radio" name="<?php echo $this->get_field_name( 'genesis_layout' ); ?>" class="default-layout" id="default-layout" value="" <?php checked( $layout, '' ); ?> /> <label class="default" for="default-layout"><?php printf( __( 'Default Layout set in <a href="%s">Theme Settings</a>', 'genesis' ), menu_page_url( 'genesis', 0 ) ); ?></label></p>

			<p><?php genesis_layout_selector( array( 'name' => $this->get_field_name( 'genesis_layout' ), 'selected' => $layout, 'type' => 'site' ) ); ?></p>
		</div>

		<br class="clear" />
		<?php
	}
	
	/**
	 * 404 Metabox
	 *
	 * @since 1.0.0
	 */
	public function metabox_404() {
		?>
		<p><?php _e( 'The <i>404 Page</i> is the page shown when the intended page was not found. You can customise it here.', 'genesis-404-page' ); ?></p>
		<p>
			<label for="<?php echo $this->get_field_id( 'title' ); ?>"><?php _e( 'Page Title:', 'genesis-404-page' ); ?></label><br />
			<input type="text" name="<?php echo $this->get_field_name( 'title' ); ?>" id="<?php echo $this->get_field_id( 'title' ); ?>" value="<?php echo esc_attr( $this->get_field_value( 'title' ) ); ?>" size="27" class="widefat" />
		</p>

		<p>
			<label for="<?php echo $this->get_field_id( 'content' ); ?>"><?php _e( 'Page Content:', 'genesis-404-page' ); ?></label><br />
			<?php wp_editor( genesis_get_option( 'content', 'genesis-404' ), 'content', array( 'textarea_name' => $this->get_field_id( 'content' ), ) ); ?>
		</p>
		<?php
	}

	/**
	 * Contextual help content.
	 *
	 * @since 1.3.0
	 */
	public function help() {

		$screen = get_current_screen();

		$help =
			'<p>'  . __( 'To customize the 404 Not Found page in your Genesis child theme, enter a title and content.', 'genesis-404-page' ) . '</p>' .
			'<p>'  . __( 'If you want to display the search form, use the shortcode [genesis-404-search].', 'genesis-404-page' ) . '</p>';

		$screen->add_help_tab( array(
			'id'      => $this->pagehook,
			'title'   => __( 'Genesis 404 Page', 'genesis' ),
			'content' => $help,
		) );

		//* Add help sidebar
		$screen->set_help_sidebar(
			'<p><strong>' . __( 'For more information:', 'genesis' ) . '</strong></p>' .
			'<p><a href="http://wordpress.org/support/plugin/genesis-404-page" target="_blank" title="' . esc_attr( __( 'Get Support', 'genesis-404-page' ) ) . '">' . __( 'Get Support', 'genesis-404-page' ) . '</a></p>' .
			'<p><a href="https://github.com/billerickson/Genesis-404-Page" target="_blank" title="' . esc_attr( __( 'Plugin Source and Development', 'genesis-404-page' ) ) . '">' . __( 'Plugin Source and Development', 'genesis-404-page' ) . '</a></p>'
		);

	}
}
